### ISSUE TITLE
Move detection problem

### ISSUE THREAD

**Issue by tosca07** [jromang/picochess#267]
Hello Jürgen,

I use picochess as well as Shredder/Arena&Fritz to play with my DGT board and only with picochess I got the following problem when executing my move on the board:

1. Start new game (picochess playing black)
2. execute first move
3. wait for answer move of picochess appearing on DGT clock display
4. Execute the move of picochess AND immediately execute new user move

=> picochess won't recognize my move (I assume because it was done too quick) saying that I should check the positions etc. I have to undo my move and execute it again to continue the game.

In blitz chess this is annoying as I will do my answer move in general very quick ;-)

First I though I might change the SLOW factor in picochess.ini but that won't help (obviously this is only a setting to correct a "sliding" move problem and not a quick answer move.

With Arena, Fritz and Shredder I don't have this problem so it is not a problem of the DGT board itself but a picochess property.

Is it possible to change somewhere the detection time of the user move which is done directly after  the picochess answer move?

Unfortunately I did not find a setting for this.

Thanks 
Dirk

P.S.
I am really enjoying playing with picochess!!!

**Comment by ScallyBag** [comment]
Hi Dirk,

Can I ask if you have if you have this setting un commented in your picochess.ini file:
disable-confirm-message = True

This can also be set via the clock under System/Display/Confirm msg/Confirm off

Cheers,

Al.

**Comment by djdekker** [comment]
If I understand correctly this issue (which I am also experiencing) has to do with the values inside the formulas of start_field_timer in dgt/board.py:

```
        if self.low_time:
            wait = (0.2 if self.channel == 'BT' else 0.10) + 0.06 * self.field_factor  # bullet => allow more sliding
        else:
            wait = (0.5 if self.channel == 'BT' else 0.25) + 0.03 * self.field_factor  # BT's scanning in half speed
```

Lower values will decrease the detection time but may have unwanted consequences which I do not know exactly (sliding problems?).

**Comment by tosca07** [comment]
Thanks for your comments.

@ScallyBag hi al - confirm msg is off (but that was a good hint nevertheless I think having this option set to on would make this problem even bigger)

@djdekker that looks better than the slow settings in picochess.ini. But strange that it is mentioned sliding problem because it should be easy to distinguish sliding problem from the problem to detect the user move after having done the picochess move.

As I am not a developer I don’t know if we can change the values there. Is the .py like a script file we can edit?

Hmmmm....

**Comment by djdekker** [comment]
I'm afraid that even setting the field_timer to zero is not going to solve this for 100%.
Let's say you start a new game and play 1.e4 and the clock responds with 1...c5. As soon as you lift Black's c pawn your DGT board registers this position:
```
rnbqkbnr
pp.ppppp
........
........
....P...
........
PPPP.PPP
RNBQKBNR
```
Now suppose you drop Black's c pawn on c5 and *immediately* play 2.d4. After all, you're playing a bullet game and don't need to think at all about this move. So you *immediately* lift the White d pawn. What happens now is that you're not giving your DGT board the opportunity to scan this position (which it is waiting for):
```
rnbqkbnr
pp.ppppp
........
..p.....
....P...
........
PPPP.PPP
RNBQKBNR
```
… but it scans this position instead, with the White d pawn missing from d2:
```
rnbqkbnr
pp.ppppp
........
..p.....
....P...
........
PPP..PPP
RNBQKBNR
```
This position cannot be reached in a legal way immediately after 1.e4, so the computer complains. Now when you drop the White pawn on d4 (thus completing your move 2.d4), the board scans this:
```
rnbqkbnr
pp.ppppp
........
..p.....
...PP...
........
PPP..PPP
RNBQKBNR
```
This is *also* a position that cannot be reached immediately after 1.e4, so the computer will complain again. In fact, it will refuse to continue playing until the board registers the exact position after 1.e4 c5 or until you take back 1.e4 (i.e., in this case, put the pieces back into starting position).

From the computer's point of view the behaviour we're seeing is perfectly correct, but from the user's point of view it is not.

So what could be a solution? Increasing the scanning speed of the DGT board seems no option. I can think of two ways:
[1] Are there perhaps any delays in the communication between PicoChess and the board that can be eliminated?
[2] In case of an illegal FEN, could PicoChess perhaps check whether the illegal FEN is the result of the player moving too quickly, that is, when an illegal FEN is followed by a FEN that would be legal if the computer's previous move were considered carried out on the board?

**Comment by tosca07** [comment]
@djdekker: interesting findings - thanks for sharing.

Then I wonder how Fritz or Shredder can detect a follow up move so quickly without problems.
My guess is that they have implemented it according to the second way you suggested.

If there is really no other delay in software layer of picochess maybe the second option is really a good way to go although I understand that from picochess point of view behavior is correct (but it would be 1000 Times bester for user perspective and especially when playing blitz...

**Comment by LocutusOfPenguin** [comment]
Its abit more complex as y describe.
What doesnt work at all is , if y move e4, d4 then then c5 and picochess is right now not suitable for playing super fast games like 1/0 ...I wanted to remove the 1/0 already, but i know that dj is using it :-)

I tried to explain here the technical backgrounds but my text getting longer and longer, so i deleted it.
Basically its so: y need to get from one stage to the other, and i need to get this info send from dgt board (which can take 400ms - so long for a 1/0 game), and processed (not that long).

Thats the reason y can move as quick as y want, and trust picochess it will recognize the steps between. I need to get this e4,c5 send from dgt board, before im ready for the new d4 move!

This cycle has nothing to do with sliding, if y have "ok pico" on clock ..but the last helps y to recognize y can forward with yr game. This is multitasking the clock display / voices are done in own process.

**Comment by tosca07** [comment]
Oh that that does not sound that it can be fixed like in other chess programs (arena, fritz, shredder) working with the dgt board.

What a pity - for me this is really annoying even in regular non blitz games in case you do your answer move very fast (eg when it is forced) after the picochess chess move and this happens regularly when you used to play this way (like against humans or as I mentioned other chess software). 

Anyway - thanks for explaining.

Regards
Dirk

**Comment by tosca07** [comment]
Yep I am having a Bluetooth board (obviously they are still too slow with scanning despite the update by dgt some time ago).

First I thought my dgt board was defect because picochess would not accept my move and inhale to take it back and execute it again.

The problem just got obviously because with the other programs I use I do not get the „error“.

But now that I know I can live with that (and I wouldn’t buy a plastic broad even if it is smarter ;-)  and I can imagine that other things are more important...

So no offense - thanks again
Dirk

**Comment by djdekker** [comment]
Hi Jürgen,

Thank you for your advise. The funny thing is, I am absolutely sure that it used to work long time ago! That was when PicoChess was in the low 0.70's range, certainly before version 0.75 in the late summer of 2016.

Take this diagram as an example. The computer is White, the user is Black. The computer announces 3.exd5 and the user intends to reply with 3...cxd5. In those days in 2016 it was possible to perform 3.exd5 cxd5 *very* quickly by just taking away the White e-pawn and Black c-pawn.

```
rnbqkbnr
pp..pppp
..p.....
...p....
...PP...
........
PPP..PPP
RNBQKBNR
```
In a discussion in the Google group back in October 2016 Torben Østergreen-Johansen shared a very similar experience:

> Let us say I have white pawns on a3 and b3, and Picochess has a black pawn on a5. In earlier versions of Picochess, I could move my b3-pawn to b4, and if Picochess responded a5xb4, I could - with a single movement of my hand - remove both blacks a5-pawn and my own a3-pawn ... and Picochess would recognize that an exchange of pawns had taken place on b4!

Obviously, this speedy way of playing moves doesn't work anymore. It cannot just have been caused by hardware limitations, because I have been using the exact same (slow) Bluetooth board and pieces ever since v0.30 or so. Something must have changed in the software.

I have always thought that the change had something to do with the way PicoChess – or perhaps python-chess – finds out how one FEN can be a legal follow-up of a previous FEN. But perhaps this doesn't make sense at all... My programming knowledge is too small to have a 100% clear insight of what's going on under the PicoChess hood.

Of course, PicoChess has evolved tremendously since those days and the last thing I'd want is to revert to v0.75. But I can't help thinking that this particular feature was very, very good... :)

**Comment by LocutusOfPenguin** [comment]
There were NEVER a look ahead...It would only work, if the software seen the e4xd5 position send from board (perhaps quicker in past?). At least as long im responsible for the code or i suffer from alzheimer :-)

**Comment by djdekker** [comment]
Hi Jürgen,

I dug up my good old DGT3000 clock from the attic and ran PicoChess v0.67 which dates back to June 2016. Here is a movie that shows what I was talking about:

http://www.youtube.com/watch?v=We_2uMr0BPY

This movie shows that PicoChess v0.67 accepted "pre-moves" (like carrying out 3.Qxd4 *before* 2...cxd4) and that the player could make his move *very* shortly after carrying out the computer move. For example, at 1:24 in the video, 10.Bd2 is carried out *after* 9...Qa5 but *before* Black's move disappears from the display. This felt really very responsive when playing bullet games!

DJ

**Comment by tosca07** [comment]
Yep - that’s the behavior I would have expected...and you have even move confirmation set to on and it works.

**Comment by djdekker** [comment]
Move detection obviously changed in v0.68 of July 2016. I fired up v0.68 tonight and made a side by side video of the same moves played in v0.67 and v0.68: http://www.youtube.com/watch?v=LkLal_-AuM0 The video makes it clear that it must have been the changes in v0.68 that made the quick ways of moving and capturing pieces impossible.

The changes concern what PicoChess does when the engine announces its best move. In v0.68 and later PicoChess waits until the user carries out the move **and** it has actually received the new position from the DGT board. In v0.67 it did not necessarily have to receive the new position from the DGT board at all: I suppose PicoChess knew the position anyway because of the engine output (by calculating a new FEN string perhaps?), thus allowing the user to play his next move more quickly.

**Comment by LocutusOfPenguin** [comment]
2years after this old version and 1/2year after this issue i finally fixed this on my upcoming v0.9R even that doesnt help y at first place.

**Comment by LocutusOfPenguin** [comment]
see eMails
